src/main/java/com/codeborne/selenide/selector/ByShadowRoot.java

```diff
@@ -0,0 +1,186 @@
+      this.shadowHost = shadowHost;
+      this.target = target;
+    }
+
+    @Override
+    @CheckReturnValue
+    @Nonnull
+    public WebElement findElement(final SearchContext context) {
+      final SearchContext root = context instanceof JavascriptExecutor ? context.findElement(By.xpath("/*")) : context;
+      if (null != this.target && 0 < this.target.length()) {
+        final List<WebElement> hosts = getShadowRoots(root, shadowHost, -1);
+        final WebElement targetWebElement = getElementInsideShadowTree(hosts, target);
+        if (targetWebElement == null) {
+          throw new NoSuchElementException("The element was not found: {" + getSelectorString() + "}");
+        }
+        return targetWebElement;
+      }
+      final List<WebElement> hosts = getShadowRoots(root, shadowHost, 1);
+      if (0 == hosts.size()) {
+        throw new NoSuchElementException("The element was not found: {" + getSelectorString() + "}");
+      }
+      return hosts.get(0);
+    }
+
+    @Override
+    @CheckReturnValue
+    @Nonnull
+    public List<WebElement> findElements(final SearchContext context) {
+      final SearchContext root = context instanceof JavascriptExecutor ? context.findElement(By.xpath("/*")) : context;
+      final List<WebElement> hosts = getShadowRoots(root, shadowHost, -1);
+      if (null != this.target && 0 < this.target.length()) {
+        return getElementsInsideShadowTree(hosts, target);
+      }
+      return hosts;
+    }
+
+    private WebElement getElementInsideShadowTree(final List<WebElement> hosts, final String target) {
+      WebElement targetWebElement = null;
+      for (final WebElement host : hosts) {
+        targetWebElement = (WebElement) getJavascriptExecutor(host)
+          .executeScript("return arguments[0].shadowRoot.querySelector(arguments[1])", host, target);
+        if (targetWebElement != null) break;
+      }
+      return targetWebElement;
+    }
+
+    @SuppressWarnings("unchecked")
+    private List<WebElement> getElementsInsideShadowTree(final List<WebElement> hosts, final String target) {
+      final List<WebElement> elems = new ArrayList<WebElement>();
+      hosts.forEach(host -> {
+        elems.addAll((List<WebElement>) getJavascriptExecutor(host)
+          .executeScript("return arguments[0].shadowRoot.querySelectorAll(arguments[1])", host, target));
+      });
+      return elems;
+    }
+
+    private JavascriptExecutor getJavascriptExecutor(final SearchContext context) {
+      JavascriptExecutor jsExecutor;
+      if (context instanceof JavascriptExecutor) {
+        jsExecutor = (JavascriptExecutor) context;
+      } else {
+        jsExecutor = (JavascriptExecutor) ((WrapsDriver) context).getWrappedDriver();
+      }
+      return jsExecutor;
+    }
+
+    @SuppressWarnings("unchecked")
+    private List<WebElement> getShadowRoots(final SearchContext host, String sh, final int limit) {
+      /*
+      console.info((function(){
```

Comment:
personally, it could be hard to maintain this JS snippet in the future

Reply:
Thanks for your comment. @dstekanov
> But I decided to leave the first one because in most cases (my own analysis) there is one shadow host or one root shadow host.

I also think that is ideal.
However, I think that it may not be the case in reality.
It expects to build a deep shadow-root unknowingly when using vue.js etc.
(I would like to end the speculation story here...)

> Also if we introduce shadowRoot("host1//host4", "p") I am not quite sure if we will be able to deprecate shadowCss(...) because a lot of users can already use it. And having two methods seems a not so good idea...

OK!
If so, shadowCss has some problems.
I will raise the issue of the problem.